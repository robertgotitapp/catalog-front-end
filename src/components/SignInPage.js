import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import { Redirect } from 'react-router-dom';
import { signIn } from '../actions/users';
import { SIGN_IN_SUCCESS, SIGN_IN_FAILED } from '../utils/const';

class SignInPage extends Component {
    state = {
      username: '',
      password: '',
      toHome: false,
    }

    handleSubmit = (e) => {
      const { username, password } = this.state;

      e.preventDefault();
      this.props.signIn({ username, password }, null)
        .then((res) => {
          if (res.actionType === SIGN_IN_SUCCESS) {
            this.setState(prevState => ({ ...prevState, toHome: true }));
          }
        });
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    render() {
      const { username, password, toHome } = this.state;

      if (toHome) {
        return <Redirect to='/' />;
      }

      return (
        <div>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Username</Form.Label>
              <Form.Control type="text" name="username" onChange={this.handleChange} value={username} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Password</Form.Label>
              <Form.Control type="password" name="password" onChange={this.handleChange} value={password} />
            </Form.Group>
            <Button variant="primary" type="submit">
              Submit
            </Button>
          </Form>
        </div>
      );
    }
}

const mapDispatchtoProps = {
  signIn,
};

export default connect(null, mapDispatchtoProps)(SignInPage);
