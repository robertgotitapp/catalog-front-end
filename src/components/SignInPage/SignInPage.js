import React, { Component } from 'react';
import { connect } from 'react-redux';
import Form from 'react-bootstrap/Form';
import Button from 'react-bootstrap/Button';
import Alert from 'react-bootstrap/Alert';
import { signIn, getUserData, loadCurrentUserData } from '../../actions/users';

export class SignInPage extends Component {
    state = {
      username: '',
      password: '',
      alerts: '',
    }

    validateInput = () => {
      const {
        username, password,
      } = this.state;
      const errors = {};
      if (username.length <= 5) {
        errors.username = 'Please enter valid username';
      }
      if (password.length <= 8) {
        errors.password = 'Please enter valid password';
      }
      if (Object.keys(errors).length !== 0) {
        this.setState({
          alerts: 'Please enter valid credentails.',
        });
        return false;
      }
      return true;
    }

    handleSubmit = (e) => {
      const { username, password } = this.state;

      e.preventDefault();
      if (this.validateInput()) {
        this.props.signIn({ username, password })
          .then((res) => {
          // If request is successful, return to Home page
            if (res.statusCode) {
              this.props.getUserData()
                .then((nextRes) => {
                  if (nextRes.statusCode) {
                    this.props.loadCurrentUserData();
                    this.props.history.push('/');
                  }
                });
            } else {
              this.setState({
                username: '',
                password: '',
                alerts: res.errors.description,
              });
            }
          });
      }
    }

    handleChange = (e) => {
      this.setState({
        [e.target.name]: e.target.value,
      });
    }

    render() {
      const { username, password } = this.state;

      return (
        <div>
          <h2> Sign In</h2>
          <Form onSubmit={this.handleSubmit}>
            <Form.Group>
              <Form.Label>Username</Form.Label>
              <Form.Control type="text" name="username" onChange={this.handleChange} value={username} />
            </Form.Group>
            <Form.Group>
              <Form.Label>Password</Form.Label>
              <Form.Control type="password" name="password" onChange={this.handleChange} value={password} />
            </Form.Group>
            <Button variant="primary" name="submit" type="submit">
              Submit
            </Button>
          </Form>
          {
            this.state.alerts
            && (
            <Alert variant='danger'>
              {this.state.alerts}
            </Alert>
            )
          }
        </div>
      );
    }
}

const mapDispatchToProps = {
  signIn,
  getUserData,
  loadCurrentUserData,
};

export default connect(null, mapDispatchToProps)(SignInPage);
