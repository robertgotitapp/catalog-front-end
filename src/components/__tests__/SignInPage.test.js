import React from 'react';
import { shallow } from 'enzyme';
import Form from 'react-bootstrap/Form';
import Alert from 'react-bootstrap/Alert';
import { SignInPage } from '../SignInPage';

describe('components/SignInPage', () => {
  const props = {};
  const container = shallow(<SignInPage {...props} />);

  it('should render correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('should contain 1 FormControl for username', () => {
    expect(container.find('[name="username"]')).toHaveLength(1);
  });

  it('should contain 1 FormControl for password', () => {
    expect(container.find('[name="password"]')).toHaveLength(1);
  });

  it('should trigger state change when FormControl username got input', () => {
    const usernameInput = container.find('[name="username"]');
    usernameInput.simulate('change', { target: { name: 'username', value: 'user1' } });
    expect(container.state().username).toEqual('user1');
  });

  it('should trigger state change when FormControl password got input', () => {
    const passwordInput = container.find('[name="password"]');
    passwordInput.simulate('change', { target: { name: 'password', value: 'Password1' } });
    expect(container.state().password).toEqual('Password1');
  });

  it('should trigger handler when form is submited', () => {
    container.instance().handleSubmit = jest.fn();
    container.instance().forceUpdate();
    const form = container.find(Form);
    form.simulate('submit');
    expect(container.instance().handleSubmit).toHaveBeenCalled();
  });

  it('should trigger alert when there is error message', () => {
    container.setState({
      username: '',
      password: '',
      toHome: false,
      alerts: 'Invalid credentials',
    });
    expect(container.find(Alert)).toHaveLength(1);
  });
});
