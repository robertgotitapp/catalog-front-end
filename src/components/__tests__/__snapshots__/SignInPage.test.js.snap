// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SignInPage should render correctly 1`] = `
<div>
  <Form
    inline={false}
    onSubmit={[Function]}
  >
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Username
      </FormLabel>
      <FormControl
        name="username"
        onChange={[Function]}
        type="text"
        value=""
      />
    </FormGroup>
    <FormGroup>
      <FormLabel
        column={false}
        srOnly={false}
      >
        Password
      </FormLabel>
      <FormControl
        name="password"
        onChange={[Function]}
        type="password"
        value=""
      />
    </FormGroup>
    <Button
      active={false}
      disabled={false}
      name="submit"
      type="submit"
      variant="primary"
    >
      Submit
    </Button>
  </Form>
</div>
`;
